src/Catel.Core/Data/BoxingCache.cs

```diff
@@ -7,17 +7,79 @@
 
 namespace Catel.Data
 {
+    using System;
     using System.Collections.Generic;
+    using Catel.Reflection;
+
+    /// <summary>
+    /// Boxing cache helper.
+    /// </summary>
+    public static class BoxingCache
+    {
+        /// <summary>
+        /// Converts the specified value into a cached boxed value in case of value type to decrease memory pressure after serialization.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        /// <returns>An object representing the value.</returns>
+        public static object GetBoxedValue(object value)
+        {
+            object objectValue = value;
+
+            if (value != null)
+            {
+                var valueType = value.GetType();
+                if (valueType.IsValueTypeEx())
+                {
+                    if (valueType == typeof(Guid))
+                    {
+                        objectValue = BoxingCache<Guid>.Default.GetBoxedValue((Guid)value);
+                    }
+                    else if (valueType == typeof(bool))
+                    {
+                        objectValue = BoxingCache<bool>.Default.GetBoxedValue((bool)value);
+                    }
+                    else if (valueType == typeof(short))
+                    {
+                        objectValue = BoxingCache<short>.Default.GetBoxedValue((short)value);
+                    }
+                    else if (valueType == typeof(ushort))
+                    {
+                        objectValue = BoxingCache<ushort>.Default.GetBoxedValue((ushort)value);
+                    }
+                    else if (valueType == typeof(int))
+                    {
+                        objectValue = BoxingCache<int>.Default.GetBoxedValue((int)value);
+                    }
+                    else if (valueType == typeof(uint))
+                    {
+                        objectValue = BoxingCache<uint>.Default.GetBoxedValue((uint)value);
+                    }
+
+                    // TODO: add more value types to support
+                }
+            }
+            else
+            {
+                objectValue = value;
+            }
+
+            return objectValue;
+        }
+    }
 
     /// <summary>
     /// Caches boxed objects to minimize the memory footprint for boxed value types.
     /// </summary>
     public class BoxingCache<T>
-        where T : struct
```

Comment:
Just more of a curiosity. I see you removed the type constraint but would have it made more sense to constrain it against IConvertible? It's supported on all the platforms Catel is targeting and makes more sense than struct for what this is being used for.

Reply:
Cool, was not aware of this interface. Will investigate, I want to enforce value types.

Reply:
Yeah it's perfect for when you want to enforce only primitive types (or anything that implements IConvertible) in an generic since struct can introduce a whole can of worms (also avoids having to do ValueType or IsPrimitive checks elsewhere).

Reply:
It does support reference values as well (e.g. a string), and that's something we definitely don't want to cache). But we can enforce that simply by checking for value type first.

Reply:
And now I remember why I didn't enforce it to anything. We use this in this method:

```
        protected static object GetObjectValue<TValue>(TValue value)
        {
            object objectValue = null;

            if (typeof(TValue).IsValueTypeEx())
            {
                objectValue = BoxingCache<TValue>.Default.GetBoxedValue(value);
            }
            else
            {
                objectValue = value;
            }

            return objectValue;
        }
```

Then we would have to add the constraint there as well, but we want to support reference types there as well (and at runtime check which path to take).

Reply:
Oh right because you're checking there. Would it not be more straightforward to do this though?
`if (typeof(IConvertible).IsAssignableFrom(typeof(TValue))` rather than the extension method? Would save the `GetTypeInfo` call on UAP targets.

Reply:
Hmmm, maybe I can implement it here:

![image](https://user-images.githubusercontent.com/1246444/66715414-6f1ba880-edc3-11e9-84e4-a172b142e0cb.png)

Use `typeof(IConvertible).IsAssignableFrom(type)` for UAP. Will need to profile performance first though.

Reply:
Maybe this may be better then?

`public static bool IsValueTypeEx(this Type type)
{
    return (type is IConvertible);
}`

Since all primitive types derive from this interface that should be all you'd need. There is the obvious side effect that if someone actually goes through the trouble of implementing IConvertible it ends up returning true but I've never seen anyone actually do that.

Reply:
Hmmm, IsAssignableFrom is an extension method, I wonder whether it does this under the hood anyway:

```
        public static bool IsAssignableFromEx(this Type type, Type typeToCheck)
        {
            Argument.IsNotNull("type", type);
            Argument.IsNotNull("typeToCheck", typeToCheck);

#if UAP_DEFAULT
            return type.IsAssignableFrom(typeToCheck);
            //return type.GetTypeInfo().IsAssignableFrom(typeToCheck.GetTypeInfo());
#else
            return type.IsAssignableFrom(typeToCheck);
#endif
        }
```

Reply:
> Maybe this may be better then?
> 
> ```
> public static bool IsValueTypeEx(this Type type)
> {
>     return (type is IConvertible);
> }
> ```
> 
> Since all primitive types derive from this interface that should be all you'd need. There is the obvious side effect that if someone actually goes through the trouble of implementing IConvertible it ends up returning true but I've never seen anyone actually do that.

This definitely makes sense for UAP, but I think it's still faster to read a property value than to check for type (but not sure, and we are talking nano-optimizations here).

Reply:
Agreed, if there’s serious concerns about performance here (doubtful) dotTrace can be used later.